File: core/src/main/java/org/readutf/hermes/response/ResponseManager.java
package org.readutf.hermes.response;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.readutf.hermes.event.Listener;
import org.readutf.hermes.packet.Packet;
import org.readutf.hermes.packet.ResponsePacket;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentLinkedQueue;

public class ResponseManager {

    private static final Logger log = LoggerFactory.getLogger(ResponseManager.class);
    private final ConcurrentHashMap<Integer, ConcurrentLinkedQueue<ResponseData<?>>> pendingResponses = new ConcurrentHashMap<>();

    public <T> CompletableFuture<T> createFuture(Packet<T> packet, Class<? extends T> type) {
        if (!packet.expectsResponse()) {
            throw new IllegalArgumentException("Packet does not expect a response");
        }

        CompletableFuture<T> future = new CompletableFuture<>();
        ConcurrentLinkedQueue<ResponseData<?>> queue = pendingResponses.computeIfAbsent(
                packet.getId(), k -> new ConcurrentLinkedQueue<>());
        queue.add(new ResponseData<>((Class<T>) type, future));
        return future;
    }


    public Listener<ResponsePacket, Void> createResponseListener() {
        return (channel, packet) -> {

            log.debug("Received response packet: {}", packet);

            int originalPacketId = packet.getOriginalPacketId();
            ConcurrentLinkedQueue<ResponseData<?>> queue = pendingResponses.get(originalPacketId);
            ResponseData<?> responseData = queue != null ? queue.poll() : null;

            if (responseData != null) {
                CompletableFuture<?> future = responseData.future;
                if(packet.isError()) {
                    future.completeExceptionally(new Exception("Error in response packet for original packet ID " + originalPacketId + ": " + packet.getResponseData()));
                    return null;
                }
                if (responseData.type.isInstance(packet.getResponseData())) {
                    responseData.complete(packet);
                } else {
                    future.completeExceptionally(new ClassCastException("Response data type mismatch"));
                }
            } else if (queue != null && queue.isEmpty()) {
                pendingResponses.remove(originalPacketId);
                log.warn("Received response for packet ID {} but no pending requests found", originalPacketId);
            }
            return null;
        };
    }

    public record ResponseData<T>(Class<T> type, CompletableFuture<T> future) {

        public void complete(ResponsePacket responsePacket) {
            future.complete(type.cast(responsePacket.getResponseData()));
        }

    }


}

File: core/src/main/kotlin/org/readutf/hermes/response/ResponseManager.kt
package org.readutf.hermes.response

import com.github.michaelbull.result.Err
import com.github.michaelbull.result.Ok
import com.github.michaelbull.result.Result
import com.github.michaelbull.result.onFailure
import com.github.michaelbull.result.onSuccess
import io.github.oshai.kotlinlogging.KLogger
import io.github.oshai.kotlinlogging.KotlinLogging
import org.readutf.hermes.Packet
import org.readutf.hermes.channel.HermesChannel
import java.util.concurrent.CompletableFuture

public class ResponseManager {
    public val logger: KLogger = KotlinLogging.logger { }

    /**
     * A map of packet id to the future that will be completed when the response is received.
     */
    public val responseFutures: MutableMap<Int, CompletableFuture<ResponseDataPacket>> =
        mutableMapOf()

    internal fun sendPacketResponse(
        packet: Packet<*>,
        result: Result<Any?, Throwable>,
        hermesChannel: HermesChannel,
    ) {
        result
            .onSuccess { data ->
                hermesChannel.sendPacket(
                    ResponseDataPacket(
                        success = true,
                        response = data,
                        error = null,
                        originalId = packet.packetId,
                    ),
                )
            }.onFailure {
                logger.debug(it) { "Failed to handle packet" }
                hermesChannel.sendPacket(
                    ResponseDataPacket(
                        success = false,
                        response = null,
                        error = it.message,
                        originalId = packet.packetId,
                    ),
                )
            }
    }

    public fun getResponseFuture(packet: ResponseDataPacket): CompletableFuture<ResponseDataPacket>? =
        responseFutures.remove(packet.originalId)

    /**
     * Creates a future that can be completed when a response is received for the given packet.
     */
    public inline fun <reified T> getPacketResponseFuture(packet: Packet<T>): CompletableFuture<Result<T, Throwable>> {
        val future = CompletableFuture<ResponseDataPacket>()
        responseFutures[packet.packetId] = future

        return future.thenApply { responsePacket ->
            logger.debug { "Received response back for packet ${packet.packetId}" }

            if (responsePacket.success) {
                if (responsePacket.response is T) {
                    logger.debug { "Received back $responsePacket as ${T::class.java.simpleName}" }
                    return@thenApply Ok(responsePacket.response as T)
                } else {
                    logger.warn { "Response packet was not of type ${T::class.java.simpleName}" }
                    Err(IllegalStateException("Response packet was not of type ${T::class.java.simpleName}"))
                }
            } else {
                logger.warn { "Received error response: ${responsePacket.error}" }
                Err(IllegalStateException(responsePacket.error))
            }
        }
    }
}
